[OSDEV-1201] Added support of geo_polygon parameter for GET v1/production-locations endpoint

VadimKovalenkoSNF opened the pull request:
Fix [OSDEV-1201](https://opensupplyhub.atlassian.net/browse/OSDEV-1201)
Added support of `geo_polygon` parameter for GET `v1/production-locations` endpoint.

[OSDEV-1201]: https://opensupplyhub.atlassian.net/browse/OSDEV-1201?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

VadimKovalenkoSNF commented on src/tests/v1/test_production_locations.py:
```diff
@@ -157,3 +157,127 @@ def test_production_locations_geo_bounding_box(self):
+        os_ids = {item["os_id"] for item in result["data"]}
+        self.assertNotIn("US202309OUTSIDE", os_ids)
+
+    def test_production_locations_geo_polygon_inside(self):
+        inside_polygon = {
+            "sector": ["Retail"],
+            "address": "Inside Polygon Address",
+            "name": "Inside Polygon Location",
+            "country": {"alpha_2": "GL"},
+            "os_id": "GL202309INSIDE",
+            "coordinates": {
+                "lon": -47.0,
+                "lat": 78.0
+            },
+        }
+
+        self.open_search_client.index(
+            index=self.production_locations_index_name,
+            body=inside_polygon,
+            id=self.open_search_client.count()
+        )
+
+        self.open_search_client.indices.refresh(
+            index=self.production_locations_index_name
+        )
+
+        query = "?geo_polygon=79.318492,-39.36719&geo_polygon=79.280399,-55.39907&geo_polygon=77.57295,-55.512304&geo_polygon=77.598154,-38.396004"
+        response = requests.get(
+            f"{self.root_url}/api/v1/production-locations/{query}",
+            headers=self.basic_headers,
+        )
+
+        result = response.json()
+        self.assertIsNotNone(result['data'])
+        self.assertEqual(len(result['data']), 1)
+        self.assertEqual(result['data'][0]['os_id'], "GL202309INSIDE")
+
```
1. `lon: -47.0, lat: 78.0` is Greenland.
2. `geo_polygon=79.318492,-39.36719&geo_polygon=79.280399,-55.39907&geo_polygon=77.57295,-55.512304&geo_polygon=77.598154,-38.396004"` - is a polygon that encloses `inside_polygon` production location.
Latitude goes first in this request. Re-check.

VadimKovalenkoSNF commented on src/tests/v1/test_production_locations.py:
```diff
@@ -157,3 +157,127 @@ def test_production_locations_geo_bounding_box(self):
+            "&geo_polygon=51.5,92.0&geo_polygon=51.0,95.0&geo_polygon=50.5,98.0"
+            "&geo_polygon=50.0,101.0&geo_polygon=49.5,104.0&geo_polygon=49.0,107.0"
+            "&geo_polygon=48.5,110.0&geo_polygon=48.0,113.0&geo_polygon=47.5,116.0"
+            "&geo_polygon=47.0,119.0&geo_polygon=46.5,122.0&geo_polygon=46.0,125.0"
+            "&geo_polygon=45.5,128.0&geo_polygon=45.0,131.0&geo_polygon=44.5,134.0"
+            "&geo_polygon=44.0,137.0&geo_polygon=43.5,140.0&geo_polygon=43.0,143.0"
+            "&geo_polygon=42.5,146.0&geo_polygon=42.0,149.0&geo_polygon=41.5,152.0"
+            "&geo_polygon=41.0,155.0&geo_polygon=40.5,158.0&geo_polygon=40.0,161.0"
+            "&geo_polygon=39.5,164.0&geo_polygon=39.0,167.0&geo_polygon=38.5,170.0"
+            "&geo_polygon=38.0,173.0&geo_polygon=37.5,176.0&geo_polygon=37.0,179.0"
+            "&geo_polygon=36.5,-178.0&geo_polygon=36.0,-175.0&geo_polygon=35.5,-172.0"
+            "&geo_polygon=35.0,-169.0&geo_polygon=34.5,-166.0&geo_polygon=34.0,-163.0"
+            "&geo_polygon=33.5,-160.0&geo_polygon=33.0,-157.0&geo_polygon=32.5,-154.0"
+            "&geo_polygon=32.0,-151.0&geo_polygon=31.5,-148.0&geo_polygon=31.0,-145.0"
+            "&geo_polygon=30.5,-142.0&geo_polygon=30.0,-139.0&geo_polygon=29.5,-136.0"
+            "&geo_polygon=29.0,-133.0&geo_polygon=28.5,-130.0&geo_polygon=28.0,-127.0"
+            "&geo_polygon=27.5,-124.0&geo_polygon=27.0,-121.0&geo_polygon=26.5,-118.0"
+            "&geo_polygon=26.0,-115.0&geo_polygon=25.5,-112.0&geo_polygon=25.0,-109.0"
+            "&geo_polygon=24.5,-106.0&geo_polygon=24.0,-103.0&geo_polygon=23.5,-100.0"
+            "&geo_polygon=23.0,-97.0&geo_polygon=22.5,-94.0&geo_polygon=22.0,-91.0"
+            "&geo_polygon=21.5,-88.0&geo_polygon=21.0,-85.0&geo_polygon=20.5,-82.0"
+            "&geo_polygon=20.0,-79.0&geo_polygon=19.5,-76.0&geo_polygon=19.0,-73.0"
+            "&geo_polygon=18.5,-70.0&geo_polygon=18.0,-67.0&geo_polygon=17.5,-64.0"
+            "&geo_polygon=17.0,-61.0&geo_polygon=16.5,-58.0&geo_polygon=16.0,-55.0"
+            "&geo_polygon=15.5,-52.0&geo_polygon=15.0,-49.0&geo_polygon=14.5,-46.0"
+            "&geo_polygon=14.0,-43.0&geo_polygon=13.5,-40.0&geo_polygon=13.0,-37.0"
+            "&geo_polygon=12.5,-34.0&geo_polygon=12.0,-31.0&geo_polygon=11.5,-28.0"
+            "&geo_polygon=11.0,-25.0&geo_polygon=10.5,-22.0&geo_polygon=10.0,-19.0"
+        )
+
+        response = requests.get(
+            f"{self.root_url}/api/v1/production-locations/{query}",
+            headers=self.basic_headers,
+        )
+
+        self.assertEqual(response.status_code, 200)
+
+        result = response.json()
+        self.assertIsNotNone(result['data'])
+        self.assertGreater(len(result['data']), 0)
```
Validatiors for incorrect `geo_polygon` parameters are tested in `src/django/api/tests/test_v1_utils.py`

protsack-stephan commented on src/django/api/views/v1/opensearch_query_builder/production_locations_query_builder.py:
```diff
@@ -140,4 +146,24 @@ def add_geo_bounding_box(self, top, left, bottom, right):
+        points = geo_polygon_filter["coordinates"]["points"]
+        for point in values:
+            lat, lon = map(float, point.split(","))
+            points.append({"lat": lat, "lon": lon})
```
Instead of mutating `geo_polygon_filter`, I would suggest to create a new version by parsing the `values` and constructing a new dictionary. We can then locate the filter in `self.query_body` and replace it with the new value. My point here is that mutating the object is much more riskier in terms of potential side effects and code clarity instead of replacing it.

VadimKovalenkoSNF replied:
Good point, thank you. I've update `add_geo_polygon()` method to construct new `geo_polygon` dict and replace old value if necessary in `self.query_body`. In a nutshell:
1. Each request a new `geo_polygon` dict is created.
2. Remove previous `geo_polygon` from `self.query_body`.
3. Append newly created `geo_polygon` to `query_body["query"]["bool"]["filter"]`.

Note, that `self.query_body["query"]["bool"]` is still mulable dict for this and other builder methods. I apply immutability only to the `geo_polygon` here.

VadimKovalenkoSNF replied:
^ It has been refactored a bit a but the main purpose remained the same - brand new `geo_polygon` wiil be created and attached to filter list.

protsack-stephan reviewed (commented):
Good work! One comment from me.

vladsha-dev commented on doc/release/RELEASE-NOTES.md:
```diff
@@ -19,7 +19,7 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
 
 ### Code/API changes
-* *Describe code/API changes here.*
+* [OSDEV-1201](https://opensupplyhub.atlassian.net/browse/OSDEV-1201) - Added support of `geo_polygon` parameter for GET `v1/production-locations` endpoint.
```
Could you please move it to the next release? I believe it won’t be included in 2.1.0.

VadimKovalenkoSNF replied:
Done.

vladsha-dev commented on src/django/api/serializers/v1/opensearch_common_validators/geo_polygon_validator.py:
```diff
@@ -0,0 +1,68 @@
+                    "field": "geo_polygon",
+                    "detail": (
+                        f"Invalid coordinate format: {point}, "
+                        "must be 'lat,lon' as floats"
```
I believe a dot should be placed at the end of the sentence.

VadimKovalenkoSNF replied:
Done.

VadimKovalenkoSNF commented on src/django/api/views/v1/opensearch_query_builder/opensearch_query_director.py:
```diff
@@ -153,3 +154,7 @@ def __process_filter(self, query_params):
                 float(top), float(left), float(bottom), float(right)
             )
+
+        geo_polygon = query_params.getlist(V1_PARAMETERS_LIST.GEO_POLYGON)
+        if geo_polygon and hasattr(self.__builder, 'add_geo_polygon'):
+            self.__builder.add_geo_polygon(geo_polygon)
```
No. `geo_bounding_box` and `geo_polygon` can belong to the same `filter`. I've tested this already, and you can check as well (locally): [link](http://localhost:6543/api/v1/production-locations/?geo_polygon=40.8,-73.7&geo_polygon=40.7,-73.5&geo_polygon=40.75,-73.3&geo_polygon=40.85,-73.4&geo_polygon=40.8,-73.7&geo_bounding_box[top]=41.3&geo_bounding_box[left]=-74.1&geo_bounding_box[bottom]=40.4&geo_bounding_box[right]=-72.3&geohex_grid_precision=10)

mazursasha1990 replied:
@VadimKovalenkoSNF You're right, it doesn't conflict. 
But I see one case that should be clarified. When drawing a custom area using the Geopolygon and then applying the Geo-bounding box that is less than custom area, the amount of found production locations will decrease. The question is this behavior expected, or should the Geopolygon be prioritized?
I think it could be clarified during the integration FE with v1 endpoints.

VadimKovalenkoSNF replied:
@mazursasha1990 

> The question is this behavior expected, or should the Geopolygon be prioritized?

I tested this and default behavior is the bounding box clips the results to only those documents inside it. Any documents outside the bounding box, even if they are inside the polygon, will not be returned.

vladsha-dev commented on src/django/api/views/v1/opensearch_query_builder/production_locations_query_builder.py:
```diff
@@ -140,4 +146,31 @@ def add_geo_bounding_box(self, top, left, bottom, right):
+            }
         }
+
+        # Remove the existing geo_polygon filter, if present,
+        # to maintain immutability.
```
Okay, but to me, this check seems like it’s being done just for the sake of checking.
If that’s not the case, it might indicate a bigger issue.
If the problem is only related to tests, then ideally, each test function shouldn’t depend on the execution of a previous one.

VadimKovalenkoSNF replied:
I've doulbe checked the builder and realized that we use `OpenSearchQueryBuilder.reset()` method each time we create a new OpenSearch query object. So this check is not necessary indeed. I removed it.

vladsha-dev replied:
Thank you, resolved.

VadimKovalenkoSNF commented on doc/release/RELEASE-NOTES.md:
```diff
@@ -19,7 +19,35 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
+* *Describe what's new here. The changes that can impact user experience should be listed in this section.*
+
+### Release instructions:
+* *Provide release instructions here.*
```
Done.

VadimKovalenkoSNF commented on doc/release/RELEASE-NOTES.md:
```diff
@@ -19,7 +19,35 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
+* *Describe migrations here.*
+
+#### Schema changes
+* *Describe schema changes here.*
```
Done.

mazursasha1990 reviewed (approved):
LGTM

mazursasha1990 reviewed (changes requested):
Hi @VadimKovalenkoSNF, after playing arround with this feature locally, I faced an unexpected error when I didn't add a value to the `geo_polygon` key.
Please take a look at it. 
As I remember, capitalizing is essential for some types of errors.
![Screenshot from 2025-04-02 18-00-07](https://github.com/user-attachments/assets/47547b70-ea1a-4c62-bd75-faaa84acffd5)

VadimKovalenkoSNF commented on src/django/api/views/v1/utils.py:
```diff
@@ -55,11 +55,25 @@ def serialize_params(serializer_class, query_params):
+                    error_response['errors'].append({
+                        'field': field,
+                        'detail': next(iter(error_data.values()))[0]
+                    })
```
Fixed.

VadimKovalenkoSNF commented:
@mazursasha1990 I've fixed the issue with handling empty value of `geo_polygon` parameter and added integration test.. Since the nature of this problem is pretty common, `serialize_params()` will now intercept similar errors for other query parameters, such as `country`, etc.

mazursasha1990 reviewed (approved):
LGTM

vladsha-dev reviewed (approved):
LGTM
